### DISCUSSION TITLE
How to add custom classes/records to reflect-config.json?

### DISCUSSION THREAD

**Discussion by jpomykala** [remkop/picocli#2110] [Q&A]
Hey there, 

I'm using Jackson for reading JSONs with Picocli and GraalVM. Unfortunately, most of JSON libraries use reflection for JSON deserialization, and as a result, I have to add them to `reflect-config.json`. I wonder if there is any way to point records/classes that would be added to the `reflect-config.json`.

I saw there is a way to provide a custom factory class instead `picocli.codegen.aot.graalvm.ReflectionConfigGenerator` but before I start doing this, I wanted to ask if there is any easier way to achieve this.

Thanks!

**Comment by yuvkun10** [comment]
Great question! There are a few approaches to add your custom classes/records to `reflect-config.json` without writing a custom factory.

**Option 1: Use the `--additional-class` flag on ReflectionConfigGenerator (simplest)**

```bash
java -cp ... picocli.codegen.aot.graalvm.ReflectionConfigGenerator \
  --additional-class=com.example.MyRecord \
  --additional-class=com.example.AnotherClass \
  com.example.MyCommand
```

This is the easiest path if you have a few extra classes to add.

**Option 2: Create a manual `reflect-config.json` and let GraalVM merge it**

Create `src/main/resources/META-INF/native-image/reflect-config.json`:

```json
[
  {
    "name": "com.example.MyRecord",
    "allDeclaredConstructors": true,
    "allDeclaredMethods": true,
    "allDeclaredFields": true
  }
]
```

GraalVM automatically merges all `reflect-config.json` files found on the classpath under `META-INF/native-image/`.

**Option 3: Use the GraalVM Tracing Agent (most thorough)**

Run your app with the tracing agent to auto-generate the config:

```bash
java -agentlib:native-image-agent=config-output-dir=src/main/resources/META-INF/native-image \
  -jar myapp.jar --your-flags
```

This generates all necessary config files including `reflect-config.json` for all classes actually used at runtime, which catches Jackson deserialization targets automatically.

The tracing agent approach is particularly good for Jackson since it exercises the actual reflection calls Jackson makes during deserialization.

See the [picocli GraalVM docs](https://picocli.info/#_graalvm_native_image) and [GraalVM reflection docs](https://www.graalvm.org/latest/reference-manual/native-image/dynamic-features/Reflection/) for full details.

Hope this helps! Let me know if you need clarification.
